File: src/pollie_watch/schema.py
"""Canonical data model shared by every ingest path and by the site.

Both houses' registers descend from the same 1984 resolutions, so they ask the
same fourteen questions in the same order. The House numbers them 1--14 on its
form; the Senate API names them. `Category` is the neutral vocabulary, and the
two mapping tables at the bottom translate each source into it.

Everything downstream (site, static API, tests) consumes `Statement` JSON
written by `Statement.model_dump_json()`, one file per person per parliament.
"""

from __future__ import annotations

from datetime import date, datetime
from enum import StrEnum
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class Chamber(StrEnum):
    HOUSE = "house"
    SENATE = "senate"


class Category(StrEnum):
    """The fourteen registrable-interest categories, in form order."""

    SHAREHOLDINGS = "shareholdings"
    TRUSTS = "trusts"
    REAL_ESTATE = "real-estate"
    DIRECTORSHIPS = "directorships"
    PARTNERSHIPS = "partnerships"
    LIABILITIES = "liabilities"
    INVESTMENTS = "investments"
    ACCOUNTS = "accounts"
    OTHER_ASSETS = "other-assets"
    OTHER_INCOME = "other-income"
    GIFTS = "gifts"
    TRAVEL_HOSPITALITY = "travel-hospitality"
    MEMBERSHIPS = "memberships"
    OTHER_INTERESTS = "other-interests"


class Holder(StrEnum):
    """Whose interest it is. The House form has a row for each; the Senate API
    does not record it, so Senate interests are `unspecified` unless the text
    says otherwise."""

    SELF = "self"
    SPOUSE = "spouse"
    DEPENDENT = "dependent"
    UNSPECIFIED = "unspecified"


class AlterationKind(StrEnum):
    ADDITION = "addition"
    DELETION = "deletion"


class ExtractionMethod(StrEnum):
    SENATE_API = "senate-api"  # structured JSON, mapped field for field
    PDF_TEXT = "pdf-text"  # text-layer PDF, tables read with pdfplumber
    OCR = "ocr"  # scanned PDF read by a local vision model, unreviewed
    MANUAL = "manual"  # a person confirmed or wrote the override file
    PENDING = "pending"  # a scan with no override yet: metadata only, no interests


# Canonical field names per category. Every `Interest.fields` dict uses only
# these keys (a subset, when a source leaves a column blank). Keep the
# per-source mapping tables below and the site's labels in step with this.
CATEGORY_FIELDS: dict[Category, tuple[str, ...]] = {
    Category.SHAREHOLDINGS: ("company",),
    Category.TRUSTS: ("name", "nature", "beneficial_interest", "role"),
    Category.REAL_ESTATE: ("location", "purpose"),
    Category.DIRECTORSHIPS: ("company", "activities"),
    Category.PARTNERSHIPS: ("name", "nature", "activities"),
    Category.LIABILITIES: ("nature", "creditor"),
    Category.INVESTMENTS: ("type", "body"),
    Category.ACCOUNTS: ("nature", "institution"),
    Category.OTHER_ASSETS: ("nature",),
    Category.OTHER_INCOME: ("nature",),
    Category.GIFTS: ("details",),
    Category.TRAVEL_HOSPITALITY: ("details",),
    Category.MEMBERSHIPS: ("organisation",),
    Category.OTHER_INTERESTS: ("nature",),
}

# Values a member writes to mean "nothing to declare". Normalised away at
# parse time so an empty category is an empty list, not a row of placeholders.
NIL_VALUES = frozenset(
    {"not applicable", "n/a", "na", "nil", "none", "nothing to declare", "-", ""}
)


def is_nil(value: str | None) -> bool:
    return value is None or value.strip().lower().rstrip(".") in NIL_VALUES


class Interest(BaseModel):
    """One declared line item, as it stands in the lodged statement."""

    model_config = ConfigDict(frozen=True)

    category: Category
    holder: Holder = Holder.UNSPECIFIED
    fields: dict[str, str] = Field(default_factory=dict)
    # Source-side identifier when there is one (Senate row GUID); lets a later
    # fetch tell an edited row from a new one.
    source_id: str | None = None

    @property
    def text(self) -> str:
        """The item flattened to one string, for search and display fallbacks."""
        return " / ".join(v for v in self.fields.values() if v)


class Alteration(BaseModel):
    """A dated addition or deletion notified after the statement was lodged.

    Both houses record these as free text against a category, so `details` is
    the record and `fields` is filled only when a parser can do so reliably.
    """

    model_config = ConfigDict(frozen=True)

    kind: AlterationKind
    category: Category | None  # None when the notification names no item
    holder: Holder = Holder.UNSPECIFIED
    details: str
    fields: dict[str, str] = Field(default_factory=dict)
    # House: the "Submitted Date" on the notification. Senate: `createdOn`.
    date_submitted: date | None = None
    # House only: "Processed by Registrar" date. The public record date.
    date_processed: date | None = None
    # Order within the source document, so a ledger can be rebuilt faithfully
    # even when two alterations share a date.
    sequence: int
    source_id: str | None = None

    @property
    def effective_date(self) -> date | None:
        """The date the public record shows: processed where the House stamps
        one, otherwise the submission date."""
        return self.date_processed or self.date_submitted


class Source(BaseModel):
    """Where a statement came from and what we fetched."""

    model_config = ConfigDict(frozen=True)

    url: str
    kind: Literal["senate-api", "house-api-pdf", "house-static-pdf"]
    sha256: str
    fetched_at: datetime
    # Path of the committed raw file, relative to the repo root.
    raw_path: str


class Extraction(BaseModel):
    model_config = ConfigDict(frozen=True)

    method: ExtractionMethod
    # Bump a parser's version when its output for the same input changes, so a
    # diff in data/ can be told apart from a change in the register.
    version: int = 1
    # Free text for anything a reader of the JSON should know: OCR model used,
    # what a reviewer corrected, a registrar's note carried over.
    notes: list[str] = Field(default_factory=list)


class Statement(BaseModel):
    """One parliamentarian's register entry for one parliament: the lodged
    statement plus every notified alteration."""

    model_config = ConfigDict(frozen=True)

    aph_id: str  # APH person ID, shared across both houses (e.g. "HWK", "316915")
    chamber: Chamber
    parliament: int
    family_name: str
    given_names: str
    display_name: str  # "Aldred, Mary" as the register lists them
    honorific: str | None = None  # "Hon", "Dr", "Senator the Hon."
    party: str | None = None
    state: str | None = None
    electorate: str | None = None  # House only
    date_lodged: date | None = None
    # The register index's own "last updated" stamp for this person.
    last_updated: date | None = None
    interests: list[Interest] = Field(default_factory=list)
    alterations: list[Alteration] = Field(default_factory=list)
    # Registrar or member notes attached to the whole statement.
    notes: list[str] = Field(default_factory=list)
    source: Source
    extraction: Extraction

    def interests_in(self, category: Category) -> list[Interest]:
        return [i for i in self.interests if i.category == category]

    def alterations_in(self, category: Category) -> list[Alteration]:
        return [a for a in self.alterations if a.category == category]


class Person(BaseModel):
    """The roster row: who someone is, independent of any one statement."""

    model_config = ConfigDict(frozen=True)

    aph_id: str
    family_name: str
    given_names: str
    preferred_name: str | None = None
    honorific: str | None = None
    post_nominals: str | None = None
    gender: str | None = None
    chamber: Chamber
    party: str | None = None
    party_code: str | None = None
    state: str | None = None
    electorate: str | None = None
    # Which parliaments we hold a statement for, in either chamber.
    parliaments: list[int] = Field(default_factory=list)
    aph_url: str | None = None


# --- Source vocabularies -----------------------------------------------------

# House form item number -> category. Item 2 has two sub-tables (beneficial
# interest, trustee) that both map to TRUSTS with `role` distinguishing them.
HOUSE_ITEMS: dict[int, Category] = {
    1: Category.SHAREHOLDINGS,
    2: Category.TRUSTS,
    3: Category.REAL_ESTATE,
    4: Category.DIRECTORSHIPS,
    5: Category.PARTNERSHIPS,
    6: Category.LIABILITIES,
    7: Category.INVESTMENTS,
    8: Category.ACCOUNTS,
    9: Category.OTHER_ASSETS,
    10: Category.OTHER_INCOME,
    11: Category.GIFTS,
    12: Category.TRAVEL_HOSPITALITY,
    13: Category.MEMBERSHIPS,
    14: Category.OTHER_INTERESTS,
}

# House table column header (as printed on the form) -> canonical field.
HOUSE_COLUMNS: dict[str, str] = {
    "Name of company": "company",
    "Name of trust/nominee company": "name",
    "Nature of its operation": "nature",
    "Nature of operation": "nature",
    "Beneficial interests": "beneficial_interest",
    "Beneficiary of the trust": "beneficial_interest",
    "Location": "location",
    "Purpose for which owned": "purpose",
    "Activities of company": "activities",
    "Name": "name",
    "Nature of interest": "nature",
    "Activities of partnership": "activities",
    "Nature of liability": "nature",
    "Creditor": "creditor",
    "Type of investment": "type",
    "Body in which investment is held": "body",
    "Nature of account": "nature",
    "Name of bank/institution": "institution",
    "Nature of any other assets": "nature",
    "Nature of income": "nature",
    "Detail of gifts": "details",
    "Details of travel/hospitality": "details",
    "Name of organisation": "organisation",
}

# Senate API top-level key -> category.
SENATE_CATEGORIES: dict[str, Category] = {
    "shareHoldings": Category.SHAREHOLDINGS,
    "trusts": Category.TRUSTS,
    "realEstate": Category.REAL_ESTATE,
    "registeredDirectorshipsOfCompanies": Category.DIRECTORSHIPS,
    "partnerships": Category.PARTNERSHIPS,
    "liabilities": Category.LIABILITIES,
    "investments": Category.INVESTMENTS,
    "savingsOrInvestmentAccounts": Category.ACCOUNTS,
    "otherAssets": Category.OTHER_ASSETS,
    "otherIncome": Category.OTHER_INCOME,
    "gifts": Category.GIFTS,
    "sponsoredTravelOrHospitality": Category.TRAVEL_HOSPITALITY,
    "officeHolderDonating": Category.MEMBERSHIPS,
    "otherInterest": Category.OTHER_INTERESTS,
}

# Senate API interest field -> canonical field.
SENATE_FIELDS: dict[str, str] = {
    "nameOfCompany": "company",  # shareholdings, directorships; trusts remaps to "name"
    "activitiesOfCompany": "activities",
    "nature": "nature",
    "interest": "beneficial_interest",
    "type": "role",
    "location": "location",
    "purposeForWhichOwned": "purpose",
    "nameOfPartnership": "name",
    "natureOfInterest": "nature",
    "activitiesOfPartnership": "activities",
    "natureOfLiability": "nature",
    "creditor": "creditor",
    "typeOfInvestment": "type",
    "bodyOfWhichInvestmentIsHeld": "body",
    "natureOfAccount": "nature",
    "nameOfBankInstitution": "institution",
    "nameOfOtherAsset": "nature",
    "nameOfIncome": "nature",
    "detailOfGifts": "details",
    "detailOfTravelHospitality": "details",
    "nameOfOrganisation": "organisation",
}

File: site/src/lib/schema.ts
// zod mirror of src/pollie_watch/schema.py. Every file under data/ is parsed
// through these once at build time, so a drift between the Python writer and
// this reader fails the build rather than rendering blanks.
import { z } from "zod";

export const CATEGORIES = [
  "shareholdings",
  "trusts",
  "real-estate",
  "directorships",
  "partnerships",
  "liabilities",
  "investments",
  "accounts",
  "other-assets",
  "other-income",
  "gifts",
  "travel-hospitality",
  "memberships",
  "other-interests",
] as const;

export type Category = (typeof CATEGORIES)[number];

export const HOLDERS = ["self", "spouse", "dependent", "unspecified"] as const;
export type Holder = (typeof HOLDERS)[number];

export const CHAMBERS = ["house", "senate"] as const;
export type Chamber = (typeof CHAMBERS)[number];

export const EXTRACTION_METHODS = ["senate-api", "pdf-text", "ocr", "manual", "pending"] as const;
export type ExtractionMethod = (typeof EXTRACTION_METHODS)[number];

const isoDate = z.string().regex(/^\d{4}-\d{2}-\d{2}$/);

export const InterestSchema = z.object({
  category: z.enum(CATEGORIES),
  holder: z.enum(HOLDERS),
  fields: z.record(z.string(), z.string()),
  source_id: z.string().nullable(),
});
export type Interest = z.infer<typeof InterestSchema>;

export const AlterationSchema = z.object({
  kind: z.enum(["addition", "deletion"]),
  category: z.enum(CATEGORIES).nullable(),
  holder: z.enum(HOLDERS),
  details: z.string(),
  fields: z.record(z.string(), z.string()),
  date_submitted: isoDate.nullable(),
  date_processed: isoDate.nullable(),
  sequence: z.number().int(),
  source_id: z.string().nullable(),
});
export type Alteration = z.infer<typeof AlterationSchema>;

export const StatementSchema = z.object({
  aph_id: z.string(),
  chamber: z.enum(CHAMBERS),
  parliament: z.number().int(),
  family_name: z.string(),
  given_names: z.string(),
  display_name: z.string(),
  honorific: z.string().nullable(),
  party: z.string().nullable(),
  state: z.string().nullable(),
  electorate: z.string().nullable(),
  date_lodged: isoDate.nullable(),
  last_updated: isoDate.nullable(),
  interests: z.array(InterestSchema),
  alterations: z.array(AlterationSchema),
  notes: z.array(z.string()),
  source: z.object({
    url: z.string(),
    kind: z.enum(["senate-api", "house-api-pdf", "house-static-pdf"]),
    sha256: z.string(),
    fetched_at: z.string(),
    raw_path: z.string(),
  }),
  extraction: z.object({
    method: z.enum(EXTRACTION_METHODS),
    version: z.number().int(),
    notes: z.array(z.string()),
  }),
});
export type Statement = z.infer<typeof StatementSchema>;

export const PersonSchema = z.object({
  aph_id: z.string(),
  family_name: z.string(),
  given_names: z.string(),
  preferred_name: z.string().nullable(),
  honorific: z.string().nullable(),
  post_nominals: z.string().nullable(),
  gender: z.string().nullable(),
  chamber: z.enum(CHAMBERS),
  party: z.string().nullable(),
  party_code: z.string().nullable(),
  state: z.string().nullable(),
  electorate: z.string().nullable(),
  parliaments: z.array(z.number().int()),
  aph_url: z.string().nullable(),
});
export type Person = z.infer<typeof PersonSchema>;

/** The fourteen items as the form numbers and names them. `short` is the
 * rubric used in ledgers and headings; `label` is the fuller question. */
export const CATEGORY_INFO: Record<Category, { item: number; short: string; label: string }> = {
  shareholdings: {
    item: 1,
    short: "Shareholdings",
    label: "Shareholdings in public and private companies",
  },
  trusts: { item: 2, short: "Trusts", label: "Family and business trusts and nominee companies" },
  "real-estate": { item: 3, short: "Real estate", label: "Real estate, by location and purpose" },
  directorships: { item: 4, short: "Directorships", label: "Directorships of companies" },
  partnerships: { item: 5, short: "Partnerships", label: "Partnerships" },
  liabilities: { item: 6, short: "Liabilities", label: "Liabilities, and the creditor concerned" },
  investments: { item: 7, short: "Investments", label: "Bonds, debentures and like investments" },
  accounts: { item: 8, short: "Accounts", label: "Savings and investment accounts" },
  "other-assets": {
    item: 9,
    short: "Other assets",
    label: "Other assets each valued at over $7,500",
  },
  "other-income": { item: 10, short: "Other income", label: "Other substantial sources of income" },
  gifts: { item: 11, short: "Gifts", label: "Gifts" },
  "travel-hospitality": {
    item: 12,
    short: "Sponsored travel",
    label: "Sponsored travel or hospitality over $300",
  },
  memberships: {
    item: 13,
    short: "Memberships",
    label: "Membership of organisations where a conflict could arise",
  },
  "other-interests": {
    item: 14,
    short: "Other interests",
    label: "Any other interest that could give rise to a conflict",
  },
};

/** Canonical field order per category (schema.py CATEGORY_FIELDS). */
export const CATEGORY_FIELDS: Record<Category, readonly string[]> = {
  shareholdings: ["company"],
  trusts: ["name", "nature", "beneficial_interest", "role"],
  "real-estate": ["location", "purpose"],
  directorships: ["company", "activities"],
  partnerships: ["name", "nature", "activities"],
  liabilities: ["nature", "creditor"],
  investments: ["type", "body"],
  accounts: ["nature", "institution"],
  "other-assets": ["nature"],
  "other-income": ["nature"],
  gifts: ["details"],
  "travel-hospitality": ["details"],
  memberships: ["organisation"],
  "other-interests": ["nature"],
};

/** The fields present across a set of interests, in canonical order. */
export function fieldOrder(
  category: Category,
  interests: { fields: Record<string, string> }[],
): string[] {
  const present = new Set(interests.flatMap((i) => Object.keys(i.fields)));
  const ordered = CATEGORY_FIELDS[category].filter((f) => present.has(f));
  const extra = [...present].filter((f) => !CATEGORY_FIELDS[category].includes(f)).toSorted();
  return [...ordered, ...extra];
}

/** Field labels per canonical field name (schema.py CATEGORY_FIELDS). */
export const FIELD_LABELS: Record<string, string> = {
  company: "Company",
  name: "Name",
  nature: "Nature",
  beneficial_interest: "Beneficial interest",
  role: "Role",
  location: "Location",
  purpose: "Purpose",
  activities: "Activities",
  creditor: "Creditor",
  type: "Type",
  body: "Body",
  institution: "Institution",
  details: "Details",
  organisation: "Organisation",
};

export const HOLDER_LABELS: Record<Holder, string> = {
  self: "Self",
  spouse: "Spouse or partner",
  dependent: "Dependent children",
  unspecified: "",
};

export const CHAMBER_LABELS: Record<Chamber, { noun: string; member: string; plural: string }> = {
  house: { noun: "House", member: "Member", plural: "Members" },
  senate: { noun: "Senate", member: "Senator", plural: "Senators" },
};
